json-everything.net/wwwroot/md/vocabs-data-2022.md

```diff
@@ -0,0 +1,121 @@
+
+- the instance data
+- the schema data
+- external JSON data
+
+where the dereferenced values serve as input for keywords in a derived subschema.
+
+The intent for this keyword is to cover the use cases discussed across several issues in the JSON Schema specification GitHub repositories.  (A quick search for `"$data"` can readily summon these issues.)
+
+## 2. Declarations
+
+The ID for this vocabulary is `https://json-everything.net/vocabs-data-2022` (the URI to this document).
+
+A draft 2020-12 meta-schema which includes this vocabulary has been defined for convenience.  The ID for the meta-schema is `https://json-everything.net/meta/data-2022`, and it can also be found at this address.
+
+## 3. The `data` Keyword
+
+### 3.1 Syntax and Semantics
+
+The value of `data` must be an object.  The keys of the object are interpreted as JSON Schema keywords, and the values MUST be one of
+
+- JSON Pointers per [RFC 6901](https://www.rfc-editor.org/rfc/rfc6901)
+- [Relative JSON Pointers](https://json-schema.org/draft/2019-09/relative-json-pointer.html)
+- URI references per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986), potentially with JSON Pointer fragments
+
+`data` MUST NOT contain any keys which are defined by the JSON Schema Core Vocabulary.
+
+`data` operates in two phases:
+
+1. All of the values are dereferenced per sections 3.2 and 3.3.
+2. The resolved object is then interpreted as a schema (the "formed schema") which is applied to the instance at the current location.
+
+The validation and annotation results of `data` are those of the formed schema.  More detail regarding output can be found in section 3.4.
+
+### 3.2 Contextual Behavior
+
+`data` MUST be processed contextually in the same manner as the schema in which it is used.  Specifically,
+
+- URI resolution MUST be performed in the same manner as `$ref` (per section 3.3).
+- Keys that are ignored by the parent schema MUST also be ignored by the formed subschema.
+
+Implementations SHOULD validate the formed schema against the containing schema's meta-schema (as specified by `$schema`) to ensure that it is a syntactically valid schema object.
+
+### 3.3 URI Resolution
+
+The values of `data` are dereferenced in different ways depending on the format of the value.
+
+If the value is a JSON Pointer, it is resolved against the instance root.
+
+If the value is a Relative JSON Pointer, it is resolved against the instance at the location currently being evaluated.
+
+If the value is a URI reference, it must be resolved in accordance with the rules of `$ref` resolution for the relevant JSON Schema specification (e.g. [draft 2020-12, §8.2](https://json-schema.org/draft/2020-12/json-schema-core.html#name-base-uri-anchors-and-derefe)).  However, unlike `$ref` which requires that the indicated data must represent a valid schema object, a `data` reference can identify any value which is valid for the associated keyword.
```

Comment:
I just realized a significant problem, and you should totally laugh at me for this after all of my pontificating about understanding URI-references properly...

Both JSON Pointers and Relative JSON Pointers are syntactically valid URI-references (treated as paths rather than fragments — so the effect would be completely different), which means there is no way to automatically detect which is which.

I think your options are:

1. split this into two keywords (and given the relative difficulty and security concerns of instance vs external support, probably separate vocabularies as the instance one is far more likely to be implemented by others), using the schema's base URI for the external keyword
2. go back to mandating full URIs (starting with a scheme) for all non-JP/RJP values.  (Note that these are not absolute-URIs, which start with a scheme but cannot have a fragment).

Option 2 would preserve the UX you want to have, and as far as accessing schema data, would be no worse than the current situation.  I know you're against option 1, I'm just including it for completeness.

Reply:
> Both JSON Pointers and Relative JSON Pointers are syntactically valid URI-references

```
/is/this/a/uri/or/a/pointer
```

🤔 yeah, I kinda see that.

I much prefer the single keyword.  Having all of the references in one place, creating a single subschema, just makes sense from a usability standpoint.  It's probably a bit harder on an implementer, but as a schema author, I'd hate to have to manage this in two places.

I see a few cases that I'd want to cater to here.

```
/plain/json/pointer
2/relative/json/pointer
#/pointer/to/a/location/in/the/schema      (uri-encoded pointer)
#anchor-within-the-schema                  (anchor syntax)
https://explicit.ly/external/data          (absolute uri)
```

I think that each of these should be uniquely identifiable.

Reply:
> Having all of the references in one place, creating a single subschema, just makes sense from a usability standpoint.

While I see your argument, I really don't agree.  To me it's a lot easier to have two things that each have a very clear purpose and syntax.  No chance of confusing a pointer and a relative path.  Not to mention, far more likely to find support for the easier vocabulary to implement.  I don't really understand at all why combining two schema templates is problematic - there are tons of reasons we combine multiple schemas, it's a normal thing to do.
  
> I think that each of these should be uniquely identifiable.

Then you have to introduce limits on what URIs are valid, which is another headache.  I find that plus the need to duplicate the schema `$id` if you want to reference it to be bigger problems than having to use two schema templates side by side (and that's as a schema author, not a plugin writer).

But again, this is your proposed vocabulary, I'm explaining myself, not demanding that you do what I'd prefer.

Reply:
🤦 `#anchor-within-the-schema` isn't necessary to support.  It's only realistic inside a schema object, at which point, you'd just use `$ref` anyway.

---

> plus the need to duplicate the schema `$id` if you want to reference it

That's why I included `#/pointer/to/a/location/in/the/schema` as an option.  This allows you to get to values within the same schema.  If you want to reference an external schema, yeah, you'd need to use the full, absolute `$id` URI for it.

---

Just for fun, let's look at an example of multiple keywords.

```json
{
  "$schema": "https://gregsdennis.github.io/json-everything/meta/data",
  "$id": "https://examp.le/schema",
  "type": "object",
  "properties": {
    "foo": {
      "type": "integer",
      "data": {
        "minimum": "/minValue",
        "maximum": "https://examp.le/data#/maxValue",
        "multipleOf": "#/values/multiples"
      }
    },
    "minValue": {
      "type": "integer"
    }
  },
  "values": {
    "multiples": 5
  },
  "dependentRequired": {
    "foo": [ "minValue" ]
  }
}

{
  "$schema": "https://gregsdennis.github.io/json-everything/meta/data",
  "$id": "https://examp.le/schema",
  "type": "object",
  "properties": {
    "foo": {
      "type": "integer",
      "instanceData": {
        "minimum": "/minValue"
      },
      "externalData": {
        "maximum": "data#/maxValue",
        "multipleOf": "#/values/multiples"
      }
    },
    "minValue": {
      "type": "integer"
    }
  },
  "values": {
    "multiples": 5
  },
  "dependentRequired": {
    "foo": [ "minValue" ]
  }
}
```

In the first example,

- we dereference all of the locations which forms a single schema object, then we process that schema
- all of the data references are under a single keyword, meaning the author only has to look in one place for them  

In the second example,

- we dereference the locations under `instanceData` which forms a schema object, then we process that schema.  Then we dereference the locations under `externalData` which also forms a schema object, and we process that one.
- the data references are split between two keywords, meaning the author needs to be sure to look in both
- is `#/values/multiples really "external" or would we need a third keyword for schema locations (`internalData`)?

***NOTE** Yes, I realize that my "internal" example isn't great, but it gets the point across.  I expect an author would want to pull a value from some other keyword elsewhere in the schema, not just a miscellaneous value store.*

Yes, the syntax of the external reference URIs are simpler in the two-keyword case, but I don't think that benefit outweighs the convenience and readability in the first.

Reply:
> That's why I included #/pointer/to/a/location/in/the/schema as an option. This allows you to get to values within the same schema. If you want to reference an external schema, yeah, you'd need to use the full, absolute $id URI for it.

Oh, right, idk why I missed that.

> the data references are split between two keywords, meaning the author needs to be sure to look in both

I'm really not trying to be difficult, I just don't understand why that's a concern.  Clearly it is for you!  But it really makes no difference at all to me.  Although I guess if you needed to mix data sources on keywords that read adjacent annotations only... hmm... that is mostly for applicators, though, which would involve... loading a schema from within the instance (which is another security concern)?  I can't imagine why you'd use `data` for an external schema when you could just use `$ref`.

> is `#/values/multiples` really "external" or would we need a third keyword for schema locations (internalData).

It's external to the instance, but I agree that the terminology here is not great.  I'm using "external" to mean "referenced by URI", pretty much, and I realize that comes across weirdly with schema data.

> Yes, the syntax of the external reference URIs are simpler in the two-keyword case, but I don't think that benefit outweighs the convenience and readability in the first.

It doesn't.  Really, the most serious objection to me is the fact that people (which perhaps just means "me") aren't as likely to implement something that requires external data access.  But that only matters if maximum adoption is your goal.  If you're more focused on implementing the use case than seeing it implemented broadly (which is totally valid), then that concern doesn't matter much.

Reply:
> people ... aren't as likely to implement something that requires external data access

Are you suggesting that `instanceData` and `externalData` (or whatever) are created as two separate vocabs rather than two keywords in the same vocab so that they can be implemented separately?

Reply:
@gregsdennis 
> Are you suggesting that `instanceData` and `externalData` (or whatever) are created as two separate vocabs rather than two keywords in the same vocab so that they can be implemented separately?

Yes.  Normally I'd prefer grouping related functionality, but one thing we learned from `format` is that if you combine simple things with difficult things, people either don't implement it at all or half-ass the whole thing.

I'd probably make two vocabularies:
* one with `instanceData` (JP & RJP) and `schemaData` (URIs/IRIs with the current schema location's base URI/IRI for resolving relative URI/IRI-references, but it resolves the value from the schema storage just like `$ref`)
* one that's just `externalData` (URIs/IRIs, but it resolves them from some separate source, either by fetching or a data storage of some sort)

Reply:
I hear you, but I'm not prepared to make multiple vocabs for data reference functionality at this time.  I think what I have (plus some pending edits) to refine the single `data` keyword is the proper next evolution for this vocab.